platform/engineering/design-docs/2021-08-01-autonomous-deployments.md

```diff
@@ -0,0 +1,197 @@
+  - [Work Estimates](#work-estimates)
+  - [Alternatives](#alternatives)
+  - [Future Work](#future-work)
+  - [Revision History](#revision-history)
+
+## Overview
+
+The document's purpose is to outline the minimum viable product (MVP) outcome to create autonomous deployment within the vets-website repository. The end of this document contains a post-MVP plan as well.
+
+### Objective
+
+VFS teams should be able to safely and reliably deploy their frontend applications to any environment at any time. This should happen without waiting for a scheduled deployment or manual intervention by Platform or other VFS teams. Deployments should also not cause significant risk to the build or to the site.
+
+While adding the ability to autonomously deploy, VA should still maintain the benefits of shared components (i.e. VFS teams are still using shared patterns/components for standardization and quality purposes, and we don’t see wide variance from one VA.gov experience to another).
+
+VFS teams should also be able to easily revert their change on their own, as needed.
+
+#### Desired Outcomes
+
+- Increasing the production deployment rate
+- Increasing customer satisfaction
+- Decreasing support volume
+- Decreasing the average time from kickoff to production, by initiative complexity
+- Decreasing build failures
+
+### Background
+
+The current deployment model is to merge changes into the `master` branch to get changes deployed to production, staging, and dev servers. Commits to staging and dev happen with relative frequency, but deployments to production are done once a day for all teams/applications through a central platform-level action. If errors occur at the application level, teams need to request an [out-of-band deployment](https://depo-platform-documentation.scrollhelp.site/support/Deployment-Policies.1279787009.html#DeploymentPolicies-Requestingout-of-banddeploys), which needs to be approved by relevant stakeholders. If the error is not critical, the fix needs to wait a full day until the next scheduled deployment.
+
+This model has increased platform support volume, deployment failures, and decreased customer satisfaction while limiting the deployments to only a handful a week. It has also increased support volume for on call teams when out-of-band deployments are needed outside business hours.
+
+Our goal with this initiative is to allow teams to freely deploy their own code, breaking free of the current model.
+
+**We are writing this with the assumption that Github Actions work will continue to completion.**
+
+This phase of the initiative does not include
+- Programmatic isolation of the applications
+- Physical isolation of the applications
+
+### High Level Design
+
+Deployment environments will be remapped from `master` as a central release location to related environment branches. `master` will deploy to production (va.gov), `staging` will deploy to staging (staging.va.gov), and `dev` will deploy to dev (dev.va.gov).
```

Comment:
I'm not so sure about this three-branch model. I don't have a different recommendation without thinking about this more deeply, but I see the potential for those three environments to get out of sync very quickly and it will be difficult to track which version of each application is deployed where.

Whose responsibility is it going to be to keep track of which version of each app is deployed where?

Reply:
yeah, I called that out the syncing below as an issue. I think we also need to figure out if we really care that they aren't synced up and if the solution below is enough.

It would be the responsibility of the team to track their own app.

Reply:
If the multi-branch approach is deemed to be best, would it be better to have two branches instead of three? Are there situations where dev and staging should be different?

Reply:
I know, as a team, we use dev more than staging, but app teams use staging mostly. I wouldn't want to kill it off but it will certainly be used less

Reply:
I can definitely see how there might be syncing issues, but also I like that the three-branch method would give each team control over how often they deploy to dev/staging/prod. I think teams work differently in terms of how often they want to test changes internally, and how often they want to share changes with external stakeholders. 

Also being able to deploy at will to staging (and not prod) means they can use staging for demos, or just have a stakeholder test out a new feature, instead of using a review instance which is behind SOCKS.

Reply:
Being able to deploy to any environment at anytime was a core requirement of the initiative. Short of recreating review instances that mirror staging (and are public), I think this is probably how we get there. Open to any other solutions as well though.

Reply:
We will need to define and document which environments are intended for what. Not necessarily as part of this design doc, just noting.

Reply:
Also, do we intend for this new staging environment workflow to essentially replace review instances?

Reply:
I don't know if I see it replacing RIs completely, I could see it being used that way but I would encourage teams to push to staging when they need something reviewed rather than pushing all changes

Reply:
this would be a huge change to the branching model, going from [trunk based development](https://trunkbaseddevelopment.com/) to [gitflow](https://nvie.com/posts/a-successful-git-branching-model/)

what does this proposed workflow look like for multiple teams working on the same repo?

Reply:
According to the case studies in [Accelerate](https://www.amazon.com/Accelerate-Software-Performing-Technology-Organizations/dp/1942788339) book, [trunk based development](https://itrevolution.com/24-key-capabilities-to-drive-improvement-in-software-delivery/) correlates to high performing teams. 

>Trunk-based development has been shown to be a predictor of high performance in software development and delivery. It is characterized by fewer than three active branches in a code repository; branches and forks having very short lifetimes (e.g., less than a day) before being merged into master; and application teams rarely or never having “code lock” periods when no one can check in code or do pull requests due to merging conflicts, code freezes, or stabilization phases.